### DISCUSSION TITLE
Replace Collection?

### DISCUSSION THREAD

**Discussion by khrome83** [pocketbase/pocketbase#3355] [Q&A]
So here is a strange question for you....

I am trying to build a system to pull about 12 tables from a public catalog, and insert them into the database.

The catalog is updated daily, but does not ship any timestamps (no updated_at or created_at fields). 

My first thought is to create a temp table adjacent to the table I want to replace. Then I would insert the new data into that table, then rename the old table, and rename the new table to be the name of the old table.

Then i would delete the old table.

I know if this was plain sqlite, this would not be any issue. But with Pocketbase, and going through the DAO, would this have any issues. I know the collections section does not offer a rename function. I imagine I would need to do raw sql for that. 

But do you see anything that would be broken by going this direction?

The other path is something like a upsert, I don't know that every table has a unique set for each row that this will work 100% of the time or not.

Some tables are a few thousand rows, while some are 1.2 million. On disk as a CSV, all are 2MBs or less, except one that is just below 100mb. So not that much data.

Because I can't do relations with ID (dataset does not always use a integer as the id, I have a 2nd column with the datasets ID, that will be used to lookup the inserted ID to build the correct relations. So it's not a straight forward CVS insert to maintain relations.

**Comment by ganigeorgiev** [comment]
> My first thought is to create a temp table adjacent to the table I want to replace. Then I would insert the new data into that table, then rename the old table, and rename the new table to be the name of the old table.
Then i would delete the old table.
I know if this was plain sqlite, this would not be any issue.

The above should work with PocketBase as long as the final table has a `_collections` entry for it (_the name of a collection is also the name of the related records table_).
To perform raw SQL queries (or use the query builder abstraction) you could check:

- (Go) https://pocketbase.io/docs/go-database/#executing-queries
- (JS `pb_hooks`) https://pocketbase.io/docs/js-database/#executing-queries

The only problem I can think of at the moment with the table replacement approach is that the realtime subscriptions will not get triggered (they are triggered only when a record is changed through `app.Dao()` save/delete handlers).

**Reply by khrome83** [reply]
I will give this a shot. Will take me a few days. Thank you for the feedback, and I will confirm and post sample code for future folks once this is done.

**Reply by ganigeorgiev** [reply]
A slightly alternative approach could be also to truncate the existing records tables and then use sql query like:
```sql
INSERT INTO DEST_TABLE (col1, col2, ...) SELECT col1, col2, ... FROM TEMP_SRC_TABLE
```
This has the benefit that you don't have to recreate indexes and views definitions every time and the missing column values (as in your case with the lack of `created`/`updated`) will be inserted with their proper defaults (_as long as a PocketBase collection first was created_).

**Reply by khrome83** [reply]
I am not 100 percent following. 

Are you suggesting to empty the table and reinsert everything from the temp table? 

I do love the idea of not having to rebuild index's and relations.

**Reply by ganigeorgiev** [reply]
> Are you suggesting to empty the table and reinsert everything from the temp table?

Yes, as an alternative to the delete/rename approach.
In terms of efficiency I'm not sure which would be the more performant, but truncating the original table and then copying the rows from the temp one will be a lot more easier to deal with (_especially if you have custom indexes or views_).

**Reply by khrome83** [reply]
Yeah I agree with the easy part. 

I wonder if this was done in a transaction, if it would have no impact to end users. 

That is my ultimate concern is having a period of time with zero records available. Especially since it impacts relations.

**Reply by ganigeorgiev** [reply]
Yes, wrapping the truncate and insert queries in a transaction would be a good idea.
To do that with PocketBase see [Go transaction](https://pocketbase.io/docs/go-database/#transaction) / [JS transaction](https://pocketbase.io/docs/js-database/#transaction) guides.
There will be always some operational downtime (or at least writes will be blocked/queued up) but considering your CSV files size I don't think it will be that bad.

**Reply by khrome83** [reply]
Thank you. Will give that a shot. 

Yeah, the theory is that everything should be a read replica of this that users are using. So hopefully their should be no meaningful impact. 

These tables are part of a catalog, there is one table that references the "entry" point to these tables. Because I can't do a relation on a non-id field, I probably have to loop through that table, and update the relations based on the desired reference, and finding the new id, and updating the related id in the table that is not being replaced.

The other option is to use the "set_number" from the catalog as the id field in the "set" table that will be replaced, something like `12312-DJASDJA-12`, but I kinda hate that. 

I am assuming collections can have IDs controlled by the user, and the auto-gen ids are optional?

Otherwise, I probably have to loop through the persistant data, pull out the row id, of that data, and the "set_number" that row should reference, and then restitch those relations after the fact.

**Reply by ganigeorgiev** [reply]
> I am assuming collections can have IDs controlled by the user, and the auto-gen ids are optional?

Yes, when inserting data with plain SQL queries you have full control over the `id` string.
